src/highlight/HighlightAnnotations.tsx

```diff
@@ -1,45 +1,170 @@
+import { AnnotationHighlight, Rect } from '../@types';
+import { CreatorHighlight, CreatorStatus, Mode } from '../store';
 import './HighlightAnnotations.scss';
 
+type CreateArg = {
+    location: number;
+    message: string;
+    shapes: Rect[];
+    text?: string;
+};
+
 type Props = {
     activeAnnotationId: string | null;
     annotations: AnnotationHighlight[];
+    createHighlight?: (arg: CreateArg) => void;
     isCreating: boolean;
+    location: number;
+    message: string;
     setActiveAnnotationId: (annotationId: string | null) => void;
+    setMessage: (message: string) => void;
+    setMode: (mode: Mode) => void;
+    setStaged: (staged: CreatorHighlight | null) => void;
+    setStatus: (status: CreatorStatus) => void;
+    staged?: CreatorHighlight | null;
+    status: CreatorStatus;
 };
 
-export default class HighlightAnnotations extends React.PureComponent<Props> {
+type State = {
+    highlightRef?: HTMLAnchorElement;
+};
+
+export function getStagedAnnotation(
+    staged?: CreatorHighlight | null,
+): Pick<AnnotationHighlight, 'id' | 'target'> | null {
+    if (!staged) {
+        return null;
+    }
+
+    const { location, shapes } = staged;
+
+    return {
+        id: 'staged',
+        target: {
+            location: {
+                type: 'page',
+                value: location,
+            },
+            shapes,
+            type: 'highlight',
+        },
+    };
+}
+
+export default class HighlightAnnotations extends React.PureComponent<Props, State> {
     static defaultProps = {
         annotations: [],
         isCreating: false,
     };
 
+    state: State = {};
+
     handleAnnotationActive = (annotationId: string | null): void => {
         const { setActiveAnnotationId } = this.props;
 
         setActiveAnnotationId(annotationId);
     };
 
+    handleCancel = (): void => {
+        const { setMessage, setStaged, setStatus } = this.props;
+        setMessage('');
+        setStaged(null);
+        setStatus(CreatorStatus.init);
+    };
+
+    handleChange = (text?: string): void => {
+        const { setMessage } = this.props;
+        setMessage(text || '');
+    };
+
+    // TODO: For testing purposes only
+    handleClick = (): void => {
+        const { location, setMode, setStaged, setStatus } = this.props;
+        // Page 1 "troublesome personality" of the Van Gogh doc
+        const shapes = [
+            {
+                height: 2.0239190432,
+                width: 21.1208338959,
+                x: 38.7415889335,
+                y: 75.0991835327,
+                type: 'rect' as const,
+            },
+        ];
+
+        setMode(Mode.HIGHLIGHT);
+        setStatus(CreatorStatus.staged);
+        setStaged({ location, shapes });
+    };
+
+    handleSubmit = (): void => {
+        const { createHighlight = noop, message, staged } = this.props;
+
+        if (!staged) {
+            return;
+        }
+
+        createHighlight({ ...(staged as CreatorHighlight), message });
```

Comment:
We should avoid explicit `as` casting where possible. The `staged` prop should already be this type.